<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2

In [75]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
query_3_1 = f'''SELECT count(id) as number_of_vacancies
                FROM vacancies'''

In [5]:
pd.read_sql_query(query_3_1, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\1469115235.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_1, connection)


number_of_vacancies
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
query_3_2 = f'''SELECT count(id) as number_of_employers
                FROM employers'''

In [7]:
pd.read_sql_query(query_3_2, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\4057088587.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_2, connection)


number_of_employers
0                23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [8]:
query_3_3 = f'''SELECT count(id) as number_of_areas
                FROM areas'''

In [9]:
pd.read_sql_query(query_3_3, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\4147122770.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_3, connection)


number_of_areas
0             1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
query_3_4 = f'''SELECT count(id) as number_of_industries
                FROM industries'''

In [11]:
pd.read_sql_query(query_3_4, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\2627729322.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_3_4, connection)


number_of_industries
0                   294

***

**Вывод:** \
В результате предварительного ознакомления с данными нам удалось поверхностно \
обрисовать картину рынка спроса на услуги IT-специалистов на платформе \
HeadHunter: всего 23501 работодатель, представляющие 294 сферы \
деятельности и 1362 региона, разместили на платформе 49197 вакансий.


# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
query_4_1 = f'''SELECT a.name AS area, count(v.id) AS cnt
                FROM vacancies AS v
                INNER JOIN areas AS a ON v.area_id = a.id
                GROUP BY a.id
                ORDER BY 2 DESC
                LIMIT 10'''

In [13]:
pd.read_sql_query(query_4_1, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\4222419522.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_1, connection)


area   cnt
0           Москва  5333
1  Санкт-Петербург  2851
2            Минск  2112
3      Новосибирск  2006
4           Алматы  1892
5     Екатеринбург  1698
6  Нижний Новгород  1670
7           Казань  1415
8        Краснодар  1301
9           Самара  1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
query_4_2 = f'''SELECT count(id)
                FROM vacancies
                WHERE salary_from IS NOT NULL OR salary_to IS NOT NULL'''

In [15]:
pd.read_sql_query(query_4_2, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\614600247.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
query_4_3 = f'''SELECT round(avg(salary_from)) AS avg_salary_from,
                       round(avg(salary_to)) AS avg_salary_to 
                FROM vacancies'''

In [17]:
pd.read_sql_query(query_4_3, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\3534961570.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_3, connection)


avg_salary_from  avg_salary_to
0          71065.0       110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
query_4_4 = f'''SELECT schedule, employment, count(id)
                FROM vacancies
                GROUP BY 1, 2
                ORDER BY 3 DESC'''

In [19]:
pd.read_sql_query(query_4_4, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\2630570843.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_4, connection)


schedule           employment  count
0        Полный день     Полная занятость  35367
1   Удаленная работа     Полная занятость   7802
2      Гибкий график     Полная занятость   1593
3   Удаленная работа  Частичная занятость   1312
4     Сменный график     Полная занятость    940
5        Полный день           Стажировка    569
6     Вахтовый метод     Полная занятость    367
7        Полный день  Частичная занятость    347
8      Гибкий график  Частичная занятость    312
9        Полный день     Проектная работа    141
10  Удаленная работа     Проектная работа    133
11     Гибкий график           Стажировка    116
12    Сменный график  Частичная занятость    101
13  Удаленная работа           Стажировка     64
14     Гибкий график     Проектная работа     18
15    Сменный график           Стажировка     12
16    Вахтовый метод     Проектная работа      2
17    Сменный график     Проектная работа      1

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
query_4_5 = f'''SELECT experience, count(id)
                FROM vacancies
                GROUP BY 1
                ORDER BY 2'''

In [21]:
pd.read_sql_query(query_4_5, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\3190483664.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_4_5, connection)


experience  count
0         Более 6 лет   1337
1           Нет опыта   7197
2       От 3 до 6 лет  14511
3  От 1 года до 3 лет  26152

***

**Вывод:** \
В большинстве своем, работодатели расположены в крупных городах (Москва, \
Санкт-Петербург, Минск и т.п.) и ищут работника с опытом работы от 1 до 6 лет \
на полный рабочий день и полную занятость. Зарплата, как правило, варьируется в \
диапазоне от 71065 до 110537  руб., однако стоит отметить, что лишь в половине \
размещенных вакансий указана хотя бы одна, нижняя или верхняя, граница зарплаты.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
query_5_1 = f'''SELECT e.name, count(v.id)
                FROM employers AS e
                INNER JOIN vacancies AS v ON e.id = v.employer_id
                GROUP BY e.id
                ORDER BY 2 DESC
                LIMIT 5'''

In [23]:
pd.read_sql_query(query_5_1, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\3902369145.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_1, connection)


name  count
0         Яндекс   1933
1     Ростелеком    491
2       Тинькофф    444
3           СБЕР    428
4  Газпром нефть    331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [73]:
query_5_2 = f'''SELECT a.name,
                       count(distinct e.id) AS count_employers,
                       count(distinct v.id) AS count_vacancies
                FROM areas AS a
                LEFT OUTER JOIN employers AS e ON a.id = e.area
                LEFT OUTER JOIN vacancies AS v ON a.id = v.area_id
                GROUP BY a.id
                ORDER BY 3, 2 DESC
                LIMIT 5'''

In [76]:
pd.read_sql_query(query_5_2, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\1514608624.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_2, connection)


name  count_employers  count_vacancies
0              Россия              410                0
1           Казахстан              207                0
2  Московская область               75                0
3  Краснодарский край               19                0
4            Беларусь               18                0

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [26]:
query_5_3 = f'''SELECT e.name, count(distinct v.area_id)
                FROM employers AS e
                LEFT OUTER JOIN vacancies AS v ON e.id = v.employer_id
                GROUP BY e.id
                ORDER BY 2 DESC'''

In [27]:
pd.read_sql_query(query_5_3, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\593965358.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_3, connection)


name  count
0                      Яндекс    181
1                  Ростелеком    152
2                  Спецремонт    116
3      Поляков Денис Иванович     88
4                   ООО ЕФИН      71
...                       ...    ...
23496              Но Смокинг      0
23497        МАСТЕРСКАЯ CREED      0
23498           Dzing Finance      0
23499                    КОРС      0
23500             Наука-Связь      0

[23501 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [28]:
query_5_4 = f'''SELECT count(id)
                FROM employers
                WHERE id NOT IN (SELECT employer_id
                                 FROM employers_industries)'''

In [29]:
pd.read_sql_query(query_5_4, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\1644714913.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_4, connection)


count
0   8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [30]:
query_5_5 = f'''SELECT name
                FROM employers
                WHERE id IN (SELECT employer_id
                           FROM employers_industries
                           GROUP BY employer_id
                           HAVING count(industry_id) = 4)
                ORDER BY name
                OFFSET 2 LIMIT 1'''

In [31]:
pd.read_sql_query(query_5_5, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\1169902027.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_5, connection)


name
0  2ГИС

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [32]:
query_5_6 = f'''SELECT count(e.id)
                FROM employers as e
                INNER JOIN employers_industries AS ei ON e.id = ei.employer_id
                INNER JOIN industries AS i ON ei.industry_id = i.id
                WHERE i.name = 'Разработка программного обеспечения' '''

In [33]:
pd.read_sql_query(query_5_6, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\2878008418.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_5_6, connection)


count
0   3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [34]:
# код для получения списка городов-милионников

In [35]:
# текст запроса

In [36]:
# результат запроса

***

**Вывод:** \
Как мы видим, по количеству размещенных вакансий с большим отрывом лидирует \
компания "Яндекс". Также в пятерке лидеров находятся "Ростелеком", \
"Тинькофф", "Сбер" и "Газпром нефть". \
Порядка 1/3 работодателей не указывают в вакансиях свою сферу деятельности.

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [77]:
query_6_1 = f'''SELECT count(name)
                FROM vacancies
                WHERE name ~* 'data|данн' '''

In [78]:
pd.read_sql_query(query_6_1, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\2814080296.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_1, connection)


count
0   1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [58]:
# текст запроса
query_6_2 = f'''SELECT count(name)
                FROM vacancies
                WHERE name ~* 'data scientist|data science|исследователь данных|^ml|[^a-z]ml|machine learning|машинн.*обучен'
                    AND (name ~* 'junior'
                        OR experience ~* 'нет опыта'
                        OR employment ~* 'стажировка')'''

In [59]:
# результат запроса
pd.read_sql_query(query_6_2, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\2428688723.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_2, connection)


count
0     51

**Я не имею ни малейшего понятия, почему с следующих заданиях запросы выдают \
неверные ответы (чуть-чуть ниже правильных). Я часа 2, если не больше, \
корректировал запрос так, чтобы он выдавал то значение, которое принимает \
платформа, но у меня это не получилось. Очень хотелось бы узнать правильный \
вариант запроса.**

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [61]:
query_6_3 = f'''select count(name)
                FROM vacancies
                WHERE name ~* 'data scientist|data science|исследователь данных|^ml|[^a-z]ml|machine learning|машинн.*обучен'
                    AND key_skills ~* 'sql' '''

In [62]:
pd.read_sql_query(query_6_3, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\3902879435.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_3, connection)


count
0    198

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [67]:
query_6_4 = f'''select count(name)
                FROM vacancies
                WHERE name ~* 'data scientist|data science|исследователь данных|^ml|[^a-z]ml|machine learning|машинн.*обучен'
                    AND key_skills ~* 'python' '''

In [68]:
pd.read_sql_query(query_6_4, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\2548700122.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_4, connection)


count
0    350

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [69]:
query_6_5 = f'''SELECT avg(ARRAY_LENGTH(REGEXP_SPLIT_TO_ARRAY(key_skills, '\t'), 1))
                FROM vacancies
                WHERE name ~* 'data scientist|data science|исследователь данных|^ml|[^a-z]ml|machine learning|машинн.*обучен' '''

In [70]:
pd.read_sql_query(query_6_5, connection)

C:\Users\drull\AppData\Local\Temp\ipykernel_15444\3854041181.py:2: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  pd.read_sql_query(query_6_5, connection)


avg
0  6.403893

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

***Это задание я, признаться честно, в экселе раскидал. Как это сделать чисто \
средствами SQL я не понял.***

***

**Вывод:** \
Среди рассматриваемых вакансий, подходящих для дата-сайентиста, в 351 требуется \
знание языка python и в 201 - SQL, при этом всего 51 вакансия подходит для новичка. \
В среднем работодатель предъявляет требования в рамках 6-7 обязательных \
навыков. \
Дата-сайентисту с опытом работы от 3 до 6 лет в среднем предлагают зарплату в \
районе 243115 рублей. 

# Общий вывод по проекту

Не могу сделать общий вывод по проекту. Полученная информация преставляет \
собою набор разрозненных фактов, из которых невозможно собрать полноценную \
картину рынка труда. Было бы хорошо провести дополнительные исследования, но \
по объему они, ориентировочно, будут превышать мастабы данного проекта.